### PR TITLE
#452: Fixes ghost determination in pattern declarations

### LINKED ISSUE #452: Pattern-matching assignment omits ghost check
A ghost cannot be assigned to a compiled variable. But this check is missing for the assignment statement that takes a pattern on the left.

```
function TwinPrimes(): (int, int) {
  (41, 43)
}

method Main() {
  // var p := TwinPrimes();  // uncommented, this gives an error, as expected
  var (x, y) := TwinPrimes();  // this should not be allowed, but the Resolver doesn't do the check
  print x, " ", y, "\n";
}
```

Without the check, the compiler produces malformed code:

```
$ dafny /compile:3 test.dfy
Dafny 2.3.0.10506

Dafny program verifier finished with 0 verified, 0 errors
Errors compiling program
/tmp/uq2q7gh0.0.cs(1628,71) : error CS0117: `_module.__default' does not contain a definition for `TwinPrimes'
```

  Rustan

### PR DESCRIPTION
**Pull request by davidcok**
Fixes #452.

### CODE REVIEW

File: Test/dafny0/Compilation.dfy [hunk 1]
```diff
@@ -234,7 +234,7 @@ module GhostLetExpr {
         ghost var z := bb + F();
         ghost var t0 := var y := z; z + 3;
         ghost var t1 := ghost var y := z; z + 3;
-        var t2 := ghost var y := z; aa + 3;
+        var t2; t2 := ghost var y := z; aa + 3;
```
Comment (RustanLeino) [inline Test/dafny0/Compilation.dfy:237]: Isn't this still okay as it was? Under the new rules, when `t2` is declared with an initializing expression, it is auto-ghost. Being auto-ghost means `t2` is automatically declared `ghost` if the RHS of the initializing assignment is ghost. But the expression `ghost var y := z; a + 33` is not ghost. So, I would expect that the auto-ghost `t2` will still be non-ghost. Am I understanding this correctly?

**Review by RustanLeino — CHANGES_REQUESTED** [review]
In the comments on https://github.com/dafny-lang/dafny/issues/452, there's an additional example given (where a `datatype` has a constructor with one ghost parameter and one non-ghost parameter). Is that example working after this PR?

File: Test/dafny0/Compilation.dfy (hunk 1 above)
Comment (davidcok) [reply Test/dafny0/Compilation.dfy:237]: var t2; declares t2 to be non-ghost. So writing var t2; t2 := ghost var y := z; aa + 3; checks that aa+3 is non-ghost. Writing var t2 := ghost var y := z; aa + 3; delays any incorrect error message to a use of t2 in a print statement (t2 is auto-ghost and ends up being non-ghost).

Comment (davidcok) [reply Test/dafny0/Compilation.dfy:237]: Added test cases and implementation code for the additional example

File: Test/git-issues/git-issue-452c.dfy [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+// RUN: %diff "%s.expect" "%t"
+
+datatype Record = Record(a: int, ghost b: int)
+
+function method TwinPrimes(): Record {
+  Record(41, 43)
+}
+
+method Main() {
+  var u := var Record(x, y) := TwinPrimes(); y; // x and w are non-ghost
```
Comment (RustanLeino) [inline Test/git-issues/git-issue-452c.dfy:15]:
```suggestion
  var u := var Record(x, y) := TwinPrimes(); y; // x is non-ghost
```

**Review by RustanLeino — APPROVED** [review]

**Pull request merged**